source/siliconlabs/EFM32GG/FlashPrg.c

```diff
@@ -0,0 +1,272 @@
+/* Flash OS Routines
+ * Copyright (c) 2009-2015 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/** @file FlashPrg.c */
+
+#include "FlashOS.h"        /* FlashOS Structures */
+#include "FlashPrg.h"
+
+/* Defines required by em_msc */
+#include "core_cm3.h"
+
+#define MSC_PRESENT
+#define MSC_COUNT            1
+#define MSC_BASE             (0x400C0000UL) /**< MSC base address  */
+#define BITBAND_PER_BASE     ((uint32_t) 0x42000000UL) /**< Peripheral Address Space bit-band area */
+#define BITBAND_RAM_BASE     ((uint32_t) 0x22000000UL) /**< SRAM Address Space bit-band area */
+#define PER_MEM_BASE         ((uint32_t) 0x40000000UL) /**< PER base address  */
+
+#include "efm32gg_msc.h"
+#define MSC_UNLOCK_CODE       0x1B71                                  /**< MSC unlock code */
+#define MSC                   ((MSC_TypeDef *) MSC_BASE)              /**< MSC base pointer */
+#include "em_msc.h"
+
+#define FLASH_PAGE_MASK (~(FLASH_PAGE_SIZE-1))
+
+#define MIN(a, b) ((a) < (b) ? (a) : (b))
+
+/* Define page size */
+#if defined(EFM32GG_1024) || defined(EFM32GG_512)
```

Comment:
This is defined in the flash structure. Why not use that config instead of duplication?? Seems error prone and a maintenance problem.